web_src/js/utils/filetree.ts

```diff
@@ -83,3 +83,7 @@ export function mergeChildIfOnlyOneDir(nodes: Item[]): void {
     }
   }
 }
+
+export function fileIsViewed(item: Item): boolean {
+  return item.isFile ? item.file.IsViewed : (item as DirItem).children.every(fileIsViewed);
```

Comment:
Is this algorithm performant?

Suppose the tree it list this: `a/b/c/d` and there are 100 files in the deepest dir `d`.

Then you need to call `fileIsViewed` about 4 * 100 = 400 times? The performance will become worse when there are more parent directories and more files.

Reply:
This approach should be acceptable for now, though I can't think of a more optimal algorithm at the moment. Not sure what else to try, haha.

Reply:
Because there are already some UI performance problems.

When using complex algorithm in Reactivity APIs, the problem becomes more complex and hard to handle.

I just want to make sure there the performance is overall good.

Reply:
Hmm, interesting choice with the top-down check.
I would have used a bottom-up algorithm:
When a file changes its state, the parent sets its state to `are all of my direct children viewed?`.
This can propagate recursively up to the root node.
That way there shouldn't be a performance concern unless a directory has an enormous number of children.
In that case, some slowdown is to be expected however.

Reply:
> I would have used a bottom-up algorithm:
> When a file changes its state, the parent sets its state to `are all of my direct children viewed?`.

Yup, that's also my intuition

Reply:
I've adopted a different implementation approach where child components propagate their state upwards to the parent. This helps eliminate redundant computation. Could you experts please review this approach? While my local tests confirm it achieves the same results, I'd like to get your professional opinions. @wxiaoguang @delvh